### PR TITLE
Introduce ranking filter

### PR DESCRIPTION
**Pull request by xMort**
Solves https://github.com/gooddata/gooddata-java/issues/949

### CODE REVIEW

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

**Review by peter-plochan — CHANGES_REQUESTED** [review]

**Comment by peter-plochan** [comment]
@xMort see comments, otherwise LGTM 👍

File: gooddata-java-model/src/main/java/com/gooddata/sdk/model/executeafm/afm/filter/RankingFilter.java [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+     *         will return a new copy of itself or its encapsulated objects that used URI qualifiers, otherwise the original object is returned.
+     *         The parameter must not be null.
+     *
+     * @return copy of itself with replaced qualifiers in case when some {@link IdentifierObjQualifier} were used, otherwise original object is returned.
+     *
+     * @throws IllegalArgumentException The exception is thrown when conversion for the identifier qualifier used by this ranking filter could not be
+     *         made by the provided
+     *         converter or when provided converter is null.
+     */
+    public RankingFilter withObjUriQualifiers(final ObjQualifierConverter objQualifierConverter) {
```
Comment (peter-plochan) [inline gooddata-java-model/src/main/java/com/gooddata/sdk/model/executeafm/afm/filter/RankingFilter.java:115]: I'm not sure about the use-case for this method, but it is very similar to [MeasureDefinition#withObjUriQualifiers](https://github.com/gooddata/gooddata-java/blob/da16ae63b448b40fadf08c0ecfd4f71f87d30bbc/gooddata-java-model/src/main/java/com/gooddata/sdk/model/executeafm/afm/MeasureDefinition.java#L73). We should consider to extract some interface for this in the future.

Comment (xMort) [reply gooddata-java-model/src/main/java/com/gooddata/sdk/model/executeafm/afm/filter/RankingFilter.java:115]: AFM supports identifier and uri qualifiers of metadata objects but old ReportDefinition executor supports only uri. Webapp needs some copy method for AFM sub-objects to recreate them with uri qualifiers only (after it extracts them and then translates them). Maybe we could make them more generic in future.

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

File: gooddata-java-model/src/main/java/com/gooddata/sdk/model/executeafm/afm/filter/RankingFilter.java [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+     *
+     * @throws NullPointerException thrown when required parameter is not provided.
+     */
+    @JsonCreator
+    public RankingFilter(
+            @JsonProperty("measures") final List<Qualifier> measures,
+            @JsonProperty("attributes") final List<Qualifier> attributes,
+            @JsonProperty("operator") final String operator,
+            @JsonProperty("value") final Integer value) {
+        this.measures = notNull(measures, "measures must not be null!");
```
[the remark this answers is not in the document]
Comment (xMort) [reply gooddata-java-model/src/main/java/com/gooddata/sdk/model/executeafm/afm/filter/RankingFilter.java:63]: Fixed.

File: gooddata-java-model/src/main/java/com/gooddata/sdk/model/executeafm/afm/filter/RankingFilter.java [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+     */
+    public List<Qualifier> getAttributes() {
+        return attributes;
+    }
+
+    /**
+     * @return ranking operator
+     */
+    @JsonIgnore
+    public RankingFilterOperator getOperator() {
```
[the remark this answers is not in the document]
Comment (xMort) [reply gooddata-java-model/src/main/java/com/gooddata/sdk/model/executeafm/afm/filter/RankingFilter.java:164]: I put there more comments. Again, I got inspiration from other similar classes.

File: gooddata-java-model/src/main/java/com/gooddata/sdk/model/executeafm/afm/filter/RankingFilter.java [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+        return RankingFilterOperator.of(this.operator);
+    }
+
+    @JsonProperty("operator")
+    public String getStringOperator() {
```
[the remark this answers is not in the document]
Comment (xMort) [reply gooddata-java-model/src/main/java/com/gooddata/sdk/model/executeafm/afm/filter/RankingFilter.java]: You are right, your name is better. I renamed the method.

**Review by peter-plochan — APPROVED** [review]

**Comment by gdgate** [comment]
Build succeeded (gate pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-java-merge-compliance-check/177/ : SUCCESS in 11s
- https://checklist.intgdc.com/job/compliance/job/gooddata-java-fossa-scanning-zuul-el7/17/ : SUCCESS in 10s (non-voting)

**Pull request merged**